### DISCUSSION TITLE
Examples - File Associations

### DISCUSSION THREAD

**Discussion by dev-mash** [tauri-apps/tauri#10599] [General]
I try to run the examples/file-associations and get this error. Any idea what I'm missing?
error: none of the selected packages contains these features: custom-protocol

Thanks in advance
Manuel

-------------------------------------

More details:
C:\Users\me\git\tauri\examples\file-associations>npm run tauri info

> file-associations@1.0.0 tauri
> node ../../tooling/cli/node/tauri.js info

WARNING: Only one package manager should be used, but found npm and yarn.
         Please remove unused package manager lock files, will use npm for now!

[✔] Environment
    - OS: Windows 10.0.19045 X64
    ✔ WebView2: 127.0.2651.98
    ✔ MSVC: Visual Studio Build Tools 2022
    ✔ rustc: 1.71.1 (eb26296b5 2023-08-03)
    ✔ cargo: 1.71.1 (7f1d04c00 2023-07-29)
    ✔ rustup: 1.26.0 (5af9b9484 2023-04-05)
    ✔ Rust toolchain: stable-x86_64-pc-windows-msvc (default)
    - node: 18.17.1
    - yarn: 1.22.19
    - npm: 10.8.2

[-] Packages
    - tauri [RUST]: 2.0.0-beta.0
    - tauri-build [RUST]: 2.0.0-beta.0
    - wry [RUST]: 0.35.2
    - tao [RUST]: 0.25.0
    - tauri-cli [RUST]: 1.4.0
    - @tauri-apps/api [NPM]: 1.6.0
    - @tauri-apps/cli [NPM]: 2.0.0-beta.1

[-] App
    - build-type: bundle
    - CSP: default-src 'self'
    - frontendDist: ["../index.html"]

C:\Users\me\git\tauri\examples\file-associations>npm run tauri build

> file-associations@1.0.0 tauri
> node ../../tooling/cli/node/tauri.js build

error: none of the selected packages contains these features: custom-protocol
       Error failed to build app: failed to build app

**Comment by FabianLars** [comment]
Sorry for the long delay here, can you pull the latest commit and try again? There were quite a few changes.

**Reply by dev-mash** [reply]
Thanks for your reply. I get following error when execution "cargo build":

error[E0599]: no method named `asset_protocol_scope` found for struct `AppHandle` in the current scope
  --> examples\file-associations\src-tauri\src\main.rs:23:34
   |
23 |   let asset_protocol_scope = app.asset_protocol_scope();
   |                                  ^^^^^^^^^^^^^^^^^^^^ method not found in `AppHandle`

**Reply by justDeek** [reply]
I just so happened to encounter the same error when trying to implement the file-association feature by following the example. It took me a few hours to figure out (since I'm still new to Rust, there were multiple things from the example not working as intended, and I had to rebuild and install the app every time to test the Open with... feature on Windows), but I finally managed to get it working.

[Here is a commit](https://github.com/iDecay-Studio/feathr/commit/05fafaba152baed3d215f50b033298f29e0b04c4#diff-60e7e7b327e755d5911fdcf27768e102de4d8aaf97aeba02fbbc87ca2a96aed5) of my changes, with the relevant files being ```src-tauri/src/file.rs```, ```src-tauri/src/lib.rs```, and ```shared/js/core/app.js```

> Side-notes: I had to revert to using webview.eval() to pass the resulting 'files' array to the frontend because the events emitters didn't work for some reason. But that's not related to the file-accociation logic.\
> I also moved the handle_file_associations logic to its own script to not bloat the main rust file.

I encountered the following roadblocks when trying to adapt the example logic:

- I had to uncomment the fs plugin usages in the scope handling section and comment out the asset-protocol lines (due to the same error that @dev-mash has; which is weird, because the v2 API docs state that AppHandle does have a asset_protocol_scope() function);

- when logging the output of ```maybe_file``` on Windows, the path to the file I opened with the app from the Desktop started with ```C:\\Users\\User\\Desktop\\...```, which means the logic under ```handle `file://` path urls and skip other urls``` ignored the file path, resulting in an empty files array (probably because ```url.to_file_path()``` failed);

- also the ```url::Url::``` is invalid and should just be ```Url::``` (at least in tauri v2);

- since the path already contains two backslashes as separators, the ```f.to_string_lossy().replace("\\", "\\\\")``` logic seems a bit awkward, since it doubles the backslashes even more. Also the ```format!("\"{file}\"",)``` was messing with my webview.eval() approach;

- the ```.join(",")``` seems to be specific to the example, since it results in a string when you probably want to have an array of files on the frontend and should only be concatenated there if desired;

- the whole ```tauri::WebviewWindowBuilder::new(&app, "main", Default::default())``` only works when the app>windows array in the tauri.config is empty, since it already creates a window called main, which I had to figure out the hard way (maybe add a note above that code);

___
I hope this helps in improving the example, so that other users don't have to go through this hassle ;)

**Reply by FabianLars** [reply]
Thanks for the feedback, i'll address this here already and try to improve the example soon.

> I had to uncomment the fs plugin usages in the scope handling section

Yep, of course. We just can't use the plugin in this repo's examples.

> comment out the asset-protocol lines (due to the same error that @dev-mash has; which is weird, because the v2 API docs state that AppHandle does have a asset_protocol_scope() function);

I'm not sure why it's not visibile in the docs.rs page, i thought that it does the annotations automatically, but the method is behind the `protocol-asset` feature flag.

> when logging the output of maybe_file on Windows, the path to the file I opened with the app from the Desktop started with C:\\Users\\User\\Desktop\\..., which means the logic under handle `file://` path urls and skip other urls ignored the file path, resulting in an empty files array (probably because url.to_file_path() failed);

oof, you're right. that part is indeed broken as Url::pase will parse normal paths too and i assume the author thought it didn't.

> also the url::Url:: is invalid and should just be Url:: (at least in tauri v2);

Not invalid, but not necessary anymore. `url::Url` imports the Url struct from the `url` crate (which must be in Cargo.toml for this to work). I assume your IDE added a `use tauri::Url` import statement that made just `Url` work too. This is a re-export of the url crate and was added after the example

> since the path already contains two backslashes as separators, the f.to_string_lossy().replace("\\", "\\\\") logic seems a bit awkward, since it doubles the backslashes even more

That may not actually the case, if you print paths rust will replace `\` with `\\` in the console output.

> Also the format!("\"{file}\"",) was messing with my webview.eval() approach;

🤷 - it's correct for the example at least as this is required to make it an array at the initialization_script step

> the .join(",") seems to be specific to the example, since it results in a string when you probably want to have an array of files on the frontend and should only be concatenated there if desired;

That is correct, it's also for the array.

> the whole tauri::WebviewWindowBuilder::new(&app, "main", Default::default()) only works when the app>windows array in the tauri.config is empty, since it already creates a window called main, which I had to figure out the hard way (maybe add a note above that code);

yeah... That's a common issue and was done in many examples.
